docs/notebooks/ask_tell_optimization.pct.py

```diff
@@ -0,0 +1,154 @@
+from trieste.observer import OBJECTIVE
+from trieste.space import Box
+
+from util.plotting import plot_regret
+
+np.random.seed(1234)
+tf.random.set_seed(1234)
+
+
+search_space = Box([0, 0], [1, 1])
+n_steps = 5
+
+def build_model(data, kernel_func=None):
+    """kernel_func should be a function that takes variance as a single input parameter"""
+    variance = tf.math.reduce_variance(data.observations)
+    if kernel_func is None:
+        kernel = gpflow.kernels.Matern52(variance=variance)
+    else:
+        kernel = kernel_func(variance)
+    gpr = gpflow.models.GPR(data.astuple(), kernel, noise_variance=1e-5)
+    gpflow.set_trainable(gpr.likelihood, False)
+
+    return GPflowModelConfig(**{
+        "model": gpr,
+        "optimizer": gpflow.optimizers.Scipy(),
+    })
+
+num_initial_points = 5
+initial_query_points = search_space.sample(num_initial_points)
+observer = mk_observer(scaled_branin)
+initial_data = observer(initial_query_points)
+model = build_model(initial_data)
+
+# %% [markdown]
+# ## Timing acquisition function: simple use case for Ask-Tell
+#
+# Let's say we are very concerned with the performance of the acqusition function, and want a simple way of measuring its performance over the course of the optimization. At the time of writing these lines, regular Trieste's optimizer does not provide such coustomization functionality, and this is where Ask-Tell comes in handy.
+
+# %%
+import timeit
+
+ask_tell = AskTellOptimizer(search_space, initial_data, model)
+
+for step in range(n_steps):
+    start = timeit.default_timer()
+    new_point = ask_tell.ask()
+    stop = timeit.default_timer()
+
+    print(f"Time at step {step + 1}: {stop - start}")
+
+    new_data = observer(new_point)
+    ask_tell.tell(new_data)
+
+
+# %% [markdown]
+# Once ask-tell optimization is over, you can get an optimization result object from it and perform whatever analysis you need. Just like with regular Trieste optimization interface. For instance here we will plot for each optimization step
+
+# %%
+def plot_ask_tell_regret(ask_tell_result):
+    observations = ask_tell_result.try_get_final_dataset().observations.numpy()
+    arg_min_idx = tf.squeeze(tf.argmin(observations, axis=0))
+
+    suboptimality = observations - SCALED_BRANIN_MINIMUM.numpy()
+    _, ax = plt.subplots(1, 2)
+    plot_regret(suboptimality, ax[0], num_init=num_initial_points, idx_best=arg_min_idx)
+
+    ax[0].set_yscale("log")
+    ax[0].set_ylabel("Regret")
+    ax[0].set_ylim(0.001, 100)
+    ax[0].set_xlabel("# evaluations")
+
+plot_ask_tell_regret(ask_tell.to_result())
+
+# %% [markdown]
+# ## Model switching: using only Ask part
+#
+# A slightly more complex use case. Let's suppose we want to switch between two models depending on some criteria dynamically during the optimization loop. Why would we do it? [Who knows!](https://bayesopt.github.io/papers/2014/paper13.pdf). In that case we can only use Ask part of the Ask-Tell interface.
```

Comment:
we are making it stranger than it is, model selection is a common issue that a lot of people would naturally wonder about - now that we would support more model types it would be even easier to do

Reply:
ok, will remove this weak attempt at humor